File: examples/Python/tornado_ioloop/rrworker.py
#!/usr/bin/env python

"""
synopsis:
    Request-reply service in Python
    Connects REP socket to tcp://localhost:5560
    Expects "Hello" from client, replies with "World"
    Modified for async/ioloop: Dave Kuhlman <dkuhlman(at)davekuhlman(dot)org>
usage:
    python rrworker.py
notes:
    To run this, start rrbroker.py, any number of instances of rrworker.py,
    and rrclient.py.
"""

import sys
import zmq
from functools import partial
from zmq.eventloop.future import Context
from zmq.eventloop.ioloop import IOLoop
from tornado import gen


@gen.coroutine
def run_worker(context):
    socket = context.socket(zmq.REP)
    socket.connect("tcp://localhost:5560")
    while True:
        message = yield socket.recv()
        print("Received request: %s" % message)
        message = message.decode('utf-8')
        reply = '{}, World'.format(message)
        reply = reply.encode('utf-8')
        yield socket.send(reply)
        print("Sent reply: {}".format(reply))


@gen.coroutine
def run(loop):
    context = Context()
    yield run_worker(context)


def main():
    args = sys.argv[1:]
    if len(args) != 0:
        sys.exit(__doc__)
    try:
        loop = IOLoop.current()
        loop.run_sync(partial(run, loop, ))
    except KeyboardInterrupt:
        print('\nFinished (interrupted)')


if __name__ == '__main__':
    main()

File: examples/Rust/rrworker.rs
#![crate_name = "rrworker"]

extern crate zmq;

use std::{thread, time};

fn main() {
    let context = zmq::Context::new();

    let responder = context.socket(zmq::REP).unwrap();
    assert!(responder.connect("tcp://localhost:5560").is_ok());

    loop {
        let string = responder.recv_string(0).unwrap().unwrap();
        println!("Received request: {}", string);

        thread::sleep(time::Duration::from_secs(1));

        responder.send_str(&"World", 0).unwrap();
    }
}
